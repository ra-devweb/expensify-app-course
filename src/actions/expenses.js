import database from '../firebase/firebase';

// Add expense
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

// Delete expense
export const deleteExpense = ({ id } = {}) => ({ type: 'DELETE_EXPENSE', id });

// Update expense
export const updateExpense = (id, update) => ({
  type: 'UPDATE_EXPENSE',
  id,
  update,
});

// Set expenses
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
