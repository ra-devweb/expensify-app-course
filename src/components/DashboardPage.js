import React from 'react';

import ExpensesList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const DashboardPage = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
  </div>
);

export default DashboardPage;
