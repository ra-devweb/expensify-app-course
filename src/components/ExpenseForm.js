import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onChangeDescription = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onChangeNote = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onChangeDate = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmitForm = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmitForm}>
          <div>
            <label htmlFor='description'>Description</label>
            <input
              type='text'
              id='description'
              placeholder='Description'
              value={this.state.description}
              onChange={this.onChangeDescription}
              autoFocus
            />
          </div>
          <div>
            <label htmlFor='amount'>Amount</label>
            <input
              id='amount'
              type='text'
              placeholder='Amount'
              value={this.state.amount}
              onChange={this.onChangeAmount}
            />
          </div>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onChangeDate}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <div>
            <label htmlFor='note'>Note (Optional)</label>
            <textarea
              id='note'
              placeholder='Add a note for expense'
              value={this.state.note}
              onChange={this.onChangeNote}
            ></textarea>
          </div>
          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
