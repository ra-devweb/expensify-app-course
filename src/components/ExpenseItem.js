import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h1>{description}</h1>
      </Link>
      <p>{amount}</p>
      <span>({createdAt})</span>
    </div>
  );
};

export default ExpenseItem;
