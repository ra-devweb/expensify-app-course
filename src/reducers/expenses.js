// Expense default state param
const expensifyDefaultState = [];

// Expensify reducer
export default (state = expensifyDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'DELETE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'UPDATE_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update,
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
