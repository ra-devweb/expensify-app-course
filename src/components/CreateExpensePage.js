import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class CreateExpensePage extends React.Component {
  onSubmit = ({ description, amount, createdAt, note }) => {
    this.props.addExpense({ description, amount, createdAt, note });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
