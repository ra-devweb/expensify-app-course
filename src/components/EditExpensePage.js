import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { updateExpense, deleteExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  updateExpense = (expense) => {
    this.props.updateExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  deleteExpense = () => {
    this.props.deleteExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.updateExpense}
        />
        <button onClick={this.deleteExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateExpense: (id, expense) => dispatch(updateExpense(id, expense)),
    deleteExpense: (id) => dispatch(deleteExpense({ id })),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
