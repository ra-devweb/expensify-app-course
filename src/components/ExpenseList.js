import React from 'react';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';
import ExpenseItem from './ExpenseItem';

export const ExpenseList = (props) => (
  <div className='container'>
    <div className='list-header'>
      <div className='show-for-mobile'>Expenses</div>
      <div className='show-for-desktop'>Expense</div>
      <div className='show-for-desktop'>Amount</div>
    </div>

    {props.expenses.length === 0 ? (
      <div className='list-body list-message'>
        <span>No expenses</span>
      </div>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseItem key={expense.id} {...expense} />;
      })
    )}
    <br />
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
