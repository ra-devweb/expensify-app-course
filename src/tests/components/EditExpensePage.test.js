import React from 'react';
import { shallow } from 'enzyme';

import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let updateExpense, deleteExpense, history, wrapper;

beforeEach(() => {
  updateExpense = jest.fn();
  deleteExpense = jest.fn();
  history = {
    push: jest.fn(),
  };
  wrapper = shallow(
    <EditExpensePage
      updateExpense={updateExpense}
      deleteExpense={deleteExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test('Should render Edit expense page correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle edit expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')({ description: 'New' });

  expect(history.push).toHaveBeenLastCalledWith('/');

  expect(updateExpense).toHaveBeenLastCalledWith('1', { description: 'New' });
});

test('Should handle remove expense', () => {
  wrapper.find('button').simulate('click');

  expect(history.push).toHaveBeenLastCalledWith('/');

  expect(deleteExpense).toHaveBeenLastCalledWith('1');
});
