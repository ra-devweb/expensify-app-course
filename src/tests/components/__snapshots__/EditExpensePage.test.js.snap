// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Edit expense page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 3500,
        "createdAt": 0,
        "description": "Rent",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <div
    className="container"
  >
    <button
      className="btn"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#e3242b",
          "color": "white",
          "marginBottom": "1rem",
        }
      }
    >
      Remove expense
    </button>
  </div>
</div>
`;
